smarty/sampler_smirky.py

```diff
@@ -296,15 +295,18 @@ def __init__(self, molecules, typetag, AtomORbases, AtomORdecorators,
                     self.reference_typename_dict[pid] = smirks
                     self.reference_typed_molecules[smile][tuple(indices)] = pid
 
-            self.reference_types = [[smirks, pid] for pid, smirks in self.reference_typename_dict.items()]
-
-            if not self.check_typed_molecules(self.reference_types):
-                raise Exception("Reference types in SMIRFF (%s) do not type all %ss in the molecules" % (self.SMIRFF, self.typetag))
+            # Check that all indices are typed for the reference
+            for smile, indicesList in self.IndexDict.items():
+                for indices in indicesList:
+                    if indices not in self.reference_typed_molecules[smile].keys():
+                        raise Exception("Reference types in SMIRFF (%s) do not \
+                                type all %ss in the molecules" % (self.SMIRFF, self.typetag))
```

Comment:
This does the same thing check_typed_molecules did, but without retyping the molecules using `self.reference_types`